mass_mailing_custom_unsubscribe/controllers/main.py

```diff
@@ -45,23 +45,18 @@ def mailing(self, mailing_id, email=None, res_id=None, token="", **post):
         _logger.debug(
             "Called `mailing()` with: %r",
             (mailing_id, email, res_id, token, post))
-        if res_id:
-            res_id = int(res_id)
         mailing = request.env["mail.mass_mailing"].sudo().browse(mailing_id)
         # Mass mailing list contacts are a special case because they have a
         # subscription management form
-        if mailing.mailing_model_name == 'mail.mass_mailing.contact':
+        if mailing.mailing_model_real == 'mail.mass_mailing.contact':
             result = super(CustomUnsubscribe, self).mailing(
-                mailing_id, email, res_id, **post)
-            # FIXME Remove res_id and token in version where this is merged:
-            # https://github.com/odoo/odoo/pull/14385
+                mailing_id, email, res_id, token=token, **post)
             result.qcontext.update({
-                "token": token,
-                "res_id": res_id,
                 "contacts": result.qcontext["contacts"].filtered(
                     lambda contact:
-                        not contact.list_id.not_cross_unsubscriptable or
-                        contact.list_id <= mailing.contact_list_ids
+                        not any(contact.list_ids.mapped(
+                            'not_cross_unsubscriptable')) or
+                        contact.list_ids <= mailing.contact_list_ids
```

Comment:
I think that, [given upstream basically broke this in v11](https://github.com/odoo/odoo/blob/6b0c7a33d6b749d79232082fb5716e6cf36a9967/addons/website_mass_mailing/views/unsubscribe_templates.xml#L20-L30), you can just add a new qcontext key like this:

```python
"lists": contact.list_ids | request.env["mail.mass_mailing.list"].search([["not_cross_unsubscriptable", "=", False]]),
```

Then iterate over those lists in the view and render 1 checkbox for each of them if contact is subscribed to any of them.

Then, when user checks some and unchecks others, record 1 `mail.mass_mailing.unsubscription` per contact+list combination.